File: benchmark/go/scatter_gather_mix.go
package main

import "fmt"

func main() {
	size := 131072
	rounds := int64(7)

	values := make([]int64, size)
	indices := make([]int, size)
	scratch := make([]int64, size)

	for i := 0; i < size; i++ {
		values[i] = ((int64(i)*1103515245 + 12345) % 2147483647) - 1073741824
		indices[i] = int((int64(i)*8191 + 127) % int64(size))
	}

	for round := int64(0); round < rounds; round++ {
		for i := 0; i < size; i++ {
			src := indices[i]
			left := values[src]
			right := values[(src+1)%size]
			scratch[i] = left + right + round
		}

		for i := 0; i < size; i++ {
			target := indices[int((int64(i)*13+round)%int64(size))]
			mixed := scratch[i]*31 + int64(i)*17 + round
			values[target] = mixed % 2147483647
		}
	}

	checksum := int64(0)
	for i, value := range values {
		checksum += value * ((int64(i) % 29) + 1)
	}

	fmt.Println(checksum)
}

File: benchmark/rust/scatter_gather_mix.rs
fn main() {
    let size = 131_072_usize;
    let rounds = 7_i64;

    let mut values = Vec::with_capacity(size);
    let mut indices = Vec::with_capacity(size);
    let mut scratch = vec![0_i64; size];

    for i in 0..size {
        values.push((((i as i64) * 1_103_515_245 + 12_345) % 2_147_483_647) - 1_073_741_824);
        indices.push(((i as i64 * 8_191 + 127) % size as i64) as usize);
    }

    for round in 0..rounds {
        for i in 0..size {
            let src = indices[i];
            let left = values[src];
            let right = values[(src + 1) % size];
            scratch[i] = left + right + round;
        }

        for i in 0..size {
            let target = indices[((i as i64 * 13 + round) as usize) % size];
            let mixed = scratch[i] * 31 + i as i64 * 17 + round;
            values[target] = mixed % 2_147_483_647;
        }
    }

    let mut checksum = 0_i64;
    for (i, value) in values.iter().enumerate() {
        checksum += *value * (((i as i64) % 29) + 1);
    }

    println!("{checksum}");
}
